File: cpp/libjoynr/websocket/WebSocketMessagingStub.cpp
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
#include "WebSocketMessagingStub.h"

#include <smrf/ByteArrayView.h>

#include "joynr/IWebSocketSendInterface.h"
#include "joynr/ImmutableMessage.h"
#include "joynr/exceptions/JoynrException.h"
#include "joynr/serializer/Serializer.h"

namespace joynr
{

WebSocketMessagingStub::WebSocketMessagingStub(std::shared_ptr<IWebSocketSendInterface> webSocket)
        : _webSocket(std::move(webSocket))
{
}

void WebSocketMessagingStub::transmit(
        std::shared_ptr<ImmutableMessage> message,
        const std::function<void(const exceptions::JoynrRuntimeException&)>& onFailure)
{

    if (!_webSocket->isInitialized()) {
        JOYNR_LOG_WARN(logger(),
                       "WebSocket not ready. Unable to send message {}",
                       message->toLogMessage());
        onFailure(exceptions::JoynrDelayMessageException(
                "WebSocket not ready. Unable to send message"));
        return;
    }

    if (logger().getLogLevel() == LogLevel::Debug) {
        JOYNR_LOG_DEBUG(logger(), ">>> OUTGOING >>> {}", message->getTrackingInfo());
    } else {
        JOYNR_LOG_TRACE(logger(), ">>> OUTGOING >>> {}", message->toLogMessage());
    }
    smrf::ByteArrayView serializedMessageView(message->getSerializedMessage());
    _webSocket->send(serializedMessageView, onFailure);
}

} // namespace joynr

File: java/messaging/websocket/joynr-websocket-common/src/main/java/io/joynr/messaging/websocket/WebSocketMessagingStub.java
/*
 * #%L
 * %%
 * Copyright (C) 2011 - 2017 BMW Car IT GmbH
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package io.joynr.messaging.websocket;

import java.util.concurrent.TimeUnit;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import io.joynr.exceptions.JoynrRuntimeException;
import io.joynr.messaging.FailureAction;
import io.joynr.messaging.IMessagingStub;
import io.joynr.messaging.SuccessAction;
import joynr.ImmutableMessage;
import joynr.system.RoutingTypes.Address;

public class WebSocketMessagingStub implements IMessagingStub {
    private static final Logger logger = LoggerFactory.getLogger(WebSocketMessagingStub.class);

    private JoynrWebSocketEndpoint webSocketEndpoint;

    private Address toAddress;

    public WebSocketMessagingStub(Address toAddress, JoynrWebSocketEndpoint webSocketEndpoint) {
        this.toAddress = toAddress;
        this.webSocketEndpoint = webSocketEndpoint;
    }

    @Override
    public void transmit(ImmutableMessage message, SuccessAction successAction, FailureAction failureAction) {
        if (logger.isTraceEnabled()) {
            logger.trace(">>> OUTGOING >>> {}", message);
        } else {
            logger.debug(">>> OUTGOING >>> {}", message.getTrackingInfo());
        }

        if (!message.isTtlAbsolute()) {
            throw new JoynrRuntimeException("Relative TTL not supported");
        }

        long timeout = message.getTtlMs() - System.currentTimeMillis();
        byte[] serializedMessage = message.getSerializedMessage();

        webSocketEndpoint.writeBytes(toAddress,
                                     serializedMessage,
                                     timeout,
                                     TimeUnit.MILLISECONDS,
                                     successAction,
                                     failureAction);
    }
}
